File: C/BSTreeDeletion.c
#include <stdio.h>
#include <stdlib.h>

struct BSTree
{
    int data;
    struct BSTree *left;
    struct BSTree *right;
};

struct BSTree* insert(struct BSTree* p, int val)
{
    struct BSTree *temp = NULL;
    if(p==NULL)
    {
        temp = (struct BSTree*)malloc(sizeof(struct BSTree));
        temp->data=val;
        temp->right=NULL;
        temp->left=NULL;
        p=temp;
    }
    else if(val<p->data)
    {
        p->left=insert(p->left,val);
    }
    else
    {
        p->right=insert(p->right,val);
    }
    return p;    
}

struct BSTree *minValueNode(struct BSTree* p)
{
    struct BSTree *current = p;
    while(current->left!=NULL)
    {
        current=current->left;
    }
    return current;
}

struct BSTree *delete(struct BSTree*p, int val)
{
    if(p==NULL)
    {
        printf("Node not found.");
        return p;
    }

    if(val<p->data)
    {
        p->left=delete(p->left,val);
    }
    else if(val>p->data)
    {
        p->right=delete(p->right,val);
    }
    else
    {
        struct BSTree *temp = NULL;
        if(p->left==NULL && p->right==NULL)
        {
            free(p);
            return NULL;
        }
        else if(p->left==NULL)
        {
            temp=p->right;
            free(p);
            return temp;
        }
        else if(p->right==NULL)
        {
            temp=p->left;
            free(p);
            return temp;
        }
        else
        {
            temp = minValueNode(p->right);
            p->data = temp->data;
            p->right=delete(p->right,temp->data);
            return p;
        }
        

    }
}

void inorder(struct BSTree *root)
{
    if(root!=NULL)
    {
        inorder(root->left);
        printf("%d ",root->data);
        inorder(root->right);
    }
}

int main()
{
    struct BSTree *root = NULL;
    int val,n,d;
    printf("Enter total numbers: ");
    scanf("%d",&n);
    for(int i=0;i<n;i++)
    {
        scanf("%d",&val);
        root = insert(root,val);
    }
    inorder(root);
    printf("enter number to delete: ");
    scanf("%d",&d);
    root = delete(root,d);
    inorder(root);


}

File: Java/BSTreeDeletion.java
import java.util.*;

class BSTree
{
    static class Node
    {
        int data;
        Node left,right;
        Node(int val)
        {
            data=val;
            left=null;
            right=null;
        }
    }

    static Node insert(Node p,int val)
    {
        if(p==null)
        {
            Node temp = new Node(val);
            p=temp;
        }
        else if(val>p.data)
        {
            p.right=insert(p.right,val);
        }
        else
        {
            p.left=insert(p.left,val);
        }
        return p;
    } 

    static Node minValueNode(Node p)
    {
        Node temp = p;
        while(temp.left!=null)
        {
            temp=temp.left;
        }
        return temp;
    }

    static Node delete(Node p, int val)
    {
        if(p==null)
        {
            System.out.println("Node not found.");
            return p;
        }
        else if(val==p.data)
        {
            Node temp=null;
            if(p.left==null && p.right==null)
            {
                p=null;
                return p;
            }
            else if(p.right==null)
            {
                temp=p.left;
                p=null;
                return temp;
            }
            else if(p.left==null)
            {
                temp=p.right;
                p=null;
                return temp;
            }
            else
            {
                temp=minValueNode(p.right);
                p.data=temp.data;
                p.right=delete(p.right,temp.data);
                return p;
            }
        }
        else if(val>p.data)
        {
            p.right=delete(p.right,val);
        }
        else
        {
            p.left=delete(p.left,val);
        }
        return p;

    }

    static void inorder(Node p)
    {
        if(p!=null)
        {
        inorder(p.left);
        System.out.print(p.data + " ");
        inorder(p.right);
        }
    }

    public static void main(String args[])
    {
        Scanner sc = new Scanner(System.in);
        Node root = null;
        root=insert(root,5);
        root=insert(root,3);
        root=insert(root,4);
        root=insert(root,1);
        root=insert(root,2);
        root=insert(root,6);
        System.out.print("Inorder: ");
        inorder(root);
        System.out.println();
        int s = sc.nextInt();
        root = delete(root,s);
        System.out.print("Inorder after deletion: ");
        inorder(root);
        System.out.println();

    }
}